sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,602 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_spi_device.h"
+
+#include "spi_device_regs.h"  // Generated.
+#include "sw/device/lib/base/memory.h"
+
+const uint16_t kDifSpiDeviceSramLen = SPI_DEVICE_BUFFER_SIZE_BYTES;
+
+dif_spi_device_result_t dif_spi_device_init(dif_spi_device_config_t config,
+                                            dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+  spi->base_addr = config.base_addr;
+
+  // NOTE: we do not write to any registers until performing all
+  // function argument checks, to avoid a halfway-configured SPI.
+  uint32_t device_config = 0;
+  switch (config.clock_polarity) {
+    case kDifSpiDeviceEdgePositive:
+      device_config |= 0x0 << SPI_DEVICE_CFG_CPOL;
+      break;
+    case kDifSpiDeviceEdgeNegative:
+      device_config |= 0x1 << SPI_DEVICE_CFG_CPOL;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config.data_phase) {
+    case kDifSpiDeviceEdgePositive:
+      device_config |= 0x1 << SPI_DEVICE_CFG_CPHA;
+      break;
+    case kDifSpiDeviceEdgeNegative:
+      device_config |= 0x0 << SPI_DEVICE_CFG_CPHA;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config.tx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      device_config |= 0x0 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      device_config |= 0x1 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config.rx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      device_config |= 0x0 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      device_config |= 0x1 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  // Ensure that we actually have a uint12_t.
+  if (config.rx_fifo_timeout > 0xfff) {
+    return kDifSpiDeviceResultBadArg;
+  }
+  device_config |= (config.rx_fifo_timeout & SPI_DEVICE_CFG_TIMER_V_MASK)
+                   << SPI_DEVICE_CFG_TIMER_V_OFFSET;
+
+  uint16_t rx_fifo_start = 0x0;
```

Comment:
I am not sure whether it should be left to the driver to decide where it wants the base for RX and TX fifo in SRAM. The peripheral allows to arbitrary select the base of these FIFOs.

Alternative would be to let driver set base as well as length, and add checks for error conditions - if RX and TX buffer interleave, etc...

Reply:
I thought about that, and I couldn't think of a reason why exposing that is useful. Program behavior is irrelevant of whether RX or TX comes first.

This way the code is simpler, and requires less checking.

Reply:
Sure, I see what you have done it this way, unless anyone else has justification why not to go with your solution, I am happy with it.